docs/src/cookbooks/hierarchy.md

```diff
@@ -0,0 +1,171 @@
+* [How do I access internal fields of an instance?](#how-do-i-access-internal-fields-of-an-instance)
+* [How do I make my parameters accessable from an instance?](#how-do-i-make-my-parameters-accessable-from-an-instance)
+* [How do I reuse a previously elaborated module, if my new module has the same parameterization?](#how-do-i-reuse-a-previously-elaborated-module-if-my-new-module-has-the-same-parameterization)
+
+## How do I instantiate multiple instances with the same module parameterization?
+
+Prior to this package, Chisel users relied on deduplication in FIRRTL to create multiple instances of the same module. This package introduces the following new APIs to enable multiply-instantiated modules directly in Chisel.
+
+`Definition(...)` enables elaborating a module, but does not actually instantiate that module. Instead, it returns a `Definition` class which represents that module's definition.
+
+`Instance(...)` takes a `Definition` and instantiates it, returning an `Instance` object.
+
+Modules (classes or traits) which will be used with the `Definition`/`Instance` api should be marked with the `@instantiable` annotation at the class/trait definition.
+
+To make a Module's members variables accessible from an `Instance` object, they must be annotated with the `@public` annotation. Note that this is only accessible from a Scala sense -- this is not in and of itself a mechanism for cross-module references.
+
+In the following example, use `Definition`, `Instance`, `@instantiable` and `@public` to create multiple instances of one specific parameterization of a module, `AddOne`.
+
+```scala mdoc:reset
+import chisel3._
+import chisel3.experimental.hierarchy._
+
+@instantiable
+class AddOne(width: Int) extends Module {
+  @public val in  = IO(Input(UInt(width.W)))
+  @public val out = IO(Output(UInt(width.W)))
+  out := in + 1.U
+}
+
+class AddTwo(width: Int) extends Module {
+  val in  = IO(Input(UInt(width.W)))
+  val out = IO(Output(UInt(width.W)))
+  val addOneDef = Definition(new AddOne(width))
+  val i0 = Instance(addOneDef)
+  val i1 = Instance(addOneDef)
+  i0.in := in
+  i1.in := i0.out
+  out   := i1.out
+}
+
+chisel3.stage.ChiselStage.emitVerilog(new AddTwo(10))
+```
+
+## How do I access internal fields of an instance?
+
+You can mark internal members of a class or trait marked with `@instantiable` with the `@public` annotation. The requirements are that the field is publically accessable, is a val or lazy val, and is a valid type.
+The list of valid types are:
+
+1. IsInstantiable
+2. IsLookupable
+3. Data
+4. BaseModule
+5. Iterable/Option containing a type that meets these requirements
+6. Basic type like String, Int, BigInt etc.
+
+To mark a superclass's member as `@public`, use the following pattern (shown with `val clock`).
+
+```scala mdoc:silent:reset
+import chisel3._
+import chisel3.experimental.hierarchy._
+
+@instantiable
+class MyModule extends Module {
+  @public val clock = clock
```

Comment:
Should this use an `override`? (Does it work with an `override`?)

Reply:
This is how it actually works (the macro rewrites it to make it legal Scala). I do think we should consider a better way to do it. Overriding wouldn't work (without a similar legalization rewrite) because you can't call `super` on a `val` and it would lead to null pointer exceptions when the `val` is used in the superclass.